microservices-orders/docs/index.rst

```diff
@@ -0,0 +1,578 @@
+.. _microservices-orders:
+
+.. toctree:: 
+    :maxdepth: 2
+
+Developing Streaming Applications
+=================================
+
+This self-paced tutorial provides exercises for developers to apply the basic principles of streaming applications.
+
+========
+Overview
+========
+
+This is a small microservice ecosystem built with Kafka Streams. There is a related `blog post <https://www.confluent.io/blog/building-a-microservices-ecosystem-with-kafka-streams-and-ksql/>`__  that outlines the approach used.
+
+.. figure:: images/microservices-demo.jpg
+    :alt: image
+
+Note: this is demo code, not a production system and certain elements are left for further work.
+
+Microservices
+~~~~~~~~~~~~~
+
+The example centers around an Orders Service which provides a REST interface to POST and GET Orders.
+Posting an Order creates an event in Kafka.
+This is picked up by three different validation engines (Fraud Service, Inventory Service, Order Details Service) which validate the order in parallel, emitting a PASS or FAIL based on whether each validation succeeds.
+The result of each validation is pushed through a separate topic, Order Validations, so that we retain the `single writer’ status of the Orders Service —> Orders Topic (there are several options for managing consistency in event collaboration, discussed in Ben Stopford's `book <https://www.confluent.io/designing-event-driven-systems>`__).
+The results of the various validation checks are aggregated back in the Order Service (Validation Aggregator) which then moves the order to a Validated or Failed state, based on the combined result.
+
+To allow users to GET any order, the Orders Service creates a queryable materialized view (embedded inside the Orders Service), using a state store in each instance of the service, so any Order can be requested historically. Note also that the Orders Service can be scaled out over a number of nodes, so GET requests must be routed to the correct node to get a certain key. This is handled automatically using the Interactive Queries functionality in Kafka Streams.
+
+The Orders Service also includes a blocking HTTP GET so that clients can read their own writes. In this way we bridge the synchronous, blocking paradigm of a Restful interface with the asynchronous, non-blocking processing performed server-side.
+
+Finally there is a very simple email service.
+
+.. figure:: images/microservices-exercises-combined.jpg
+    :alt: image
+
+All the services are client applications written in Java, and they use the Kafka Streams API.
+The java source code for these microservices are in the `kafka-streams-examples repo <https://github.com/confluentinc/kafka-streams-examples/tree/5.0.1-post/src/main/java/io/confluent/examples/streams/microservices>`__.
```

Comment:
A meta question: this PR is against 5.0.1-post and hence it is fine to explicitly specify the version in the url, but when we `pint-merge` it to downstream branches it means future versions will also be referring to 5.0.1-post branch only, right?

Reply:
@guozhangwang :  Thanks for pointing this out.  Ideally I would use `|release|` -- I've opened a jira DOCS-997 that I need to first resolve, and then I will be able to use `|release|` to automatically refer to current release.  This will allow `pint merge` to work seamlessly.

Reply:
I've implemented DOCS-997 and removed all hardcoded references to `5.0.1-post`